packages/svelte/src/compiler/phases/3-transform/utils.js

```diff
@@ -237,51 +238,48 @@ export function infer_namespace(namespace, parent, nodes, path) {
  * @param {import('#compiler').Namespace | 'keep' | 'maybe_html'} namespace
  */
 function check_nodes_for_namespace(nodes, namespace) {
+	/**
+	 * @param {import('#compiler').SvelteNode} node}
+	 * @param {{next: () => void}} context
+	 */
+	const SvelteBlock = (node, { next }) => next();
+
+	/**
+	 * @param {import('#compiler').SvelteElement | import('#compiler').RegularElement} node}
+	 * @param {{stop: () => void}} context
+	 */
+	const RegularElement = (node, { stop }) => {
+		if (!node.metadata.svg) {
+			namespace = 'html';
+			stop();
+		} else if (namespace === 'keep') {
+			namespace = 'svg';
+		}
+	};
+
 	for (const node of nodes) {
-		if (node.type === 'RegularElement' || node.type === 'SvelteElement') {
-			if (!node.metadata.svg) {
-				namespace = 'html';
-				break;
-			} else if (namespace === 'keep') {
-				namespace = 'svg';
-			}
-		} else if (
-			(node.type === 'Text' && node.data.trim() !== '') ||
-			node.type === 'HtmlTag' ||
-			node.type === 'RenderTag'
-		) {
-			namespace = 'maybe_html';
-		} else if (node.type === 'EachBlock') {
-			namespace = check_nodes_for_namespace(node.body.nodes, namespace);
-			if (namespace === 'html') break;
-			if (node.fallback) {
-				namespace = check_nodes_for_namespace(node.fallback.nodes, namespace);
-				if (namespace === 'html') break;
-			}
-		} else if (node.type === 'IfBlock') {
-			namespace = check_nodes_for_namespace(node.consequent.nodes, namespace);
-			if (namespace === 'html') break;
-			if (node.alternate) {
-				namespace = check_nodes_for_namespace(node.alternate.nodes, namespace);
-				if (namespace === 'html') break;
-			}
-		} else if (node.type === 'AwaitBlock') {
-			if (node.pending) {
-				namespace = check_nodes_for_namespace(node.pending.nodes, namespace);
-				if (namespace === 'html') break;
+		walk(
+			node,
+			{},
+			{
+				EachBlock: SvelteBlock,
+				IfBlock: SvelteBlock,
+				AwaitBlock: SvelteBlock,
+				KeyBlock: SvelteBlock,
```

Comment:
I think this isn't quite correct. `SvelteBlock` isn't necessary because by default nodes are traversed recursively. What we instead need is to prevent traversal into any children except those nodes we know are save to traverse - else this would traverse into a snippet for example and could yield false results.
```suggestion
				_(node, { next }) {
					if (
						node.type === 'EachBlock' ||
						node.type === 'IfBlock' ||
						node.type === 'AwaitBlock' ||
						node.type === 'Fragment' ||
						node.type === 'KeyBlock' ||
						node.type === 'RegularElement' ||
						node.type === 'SvelteElement' ||
						node.type === 'Text' ||
						node.type === 'HtmlTag' ||
						node.type === 'RenderTag'
					) {
						next();
					}
				},
```